File: backend/models/bist_ownership.py
"""
Response models for the BIST ownership board.

A parallel set to `models/ownership.py` rather than a reuse of it, and the
split is about currency and evidence, not taste. The global board is USD and
its positions are quarterly portfolio filings; this one is lira and its
positions are *stakes* — a share of one company's capital, disclosed only
once it crosses 5%. Every share here is a **fraction**, as on every other
`/api/bist/*` payload: `0.4912`, never `49.12`.

`value_try` is a marked figure (`stake × market cap`) for a shareholder and a
reported one for a fund, and `value_basis` says which, because the two are not
the same kind of number and a column that mixed them would be wrong for half
its rows.

The house rules are the same as the global board's and are worth restating:
a figure nobody published is `None` and renders as unknown, never as 0; every
position carries its own source; and the board is never handed to a page as an
empty list, because "no data" and "holds nothing" must not look alike.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

HolderCategory = Literal["holding", "state", "foreign", "fund", "other"]

SourceKind = Literal["isyatirim_shareholders", "kap_fund_report"]

ValueBasis = Literal["marked", "reported", "unknown"]

# How a stake changed between two daily snapshots of the shareholder table.
StakeMoveKind = Literal["new", "exit", "add", "trim"]

# Key of the pooled tail slice in an allocation, so the UI can style it apart
# from real holdings without matching on a label.
POOLED_SLICE_KEY = "__other__"

# The KAP filing classes that describe ownership changing hands, in the order
# they matter. The tape is classified by `kap_materiality`; this is the subset
# that becomes a "move" on this board.
OWNERSHIP_EVENTS: tuple[str, ...] = (
    "icsel_islem",
    "satin_alma",
    "pay_alim_teklifi",
    "halka_arz",
    "sermaye",
    "geri_alim",
)


class SourceRef(BaseModel):
    kind: SourceKind
    label: str
    url: str | None = None
    as_of: str | None = None
    """What date the figure describes. For a fund, the report month."""
    retrieved_at: str | None = None
    """When it was fetched. Kept apart from `as_of` on purpose."""


class Position(BaseModel):
    ticker: str
    name: str
    stake_pct: float | None = None
    """Share of the company's capital, as a fraction. `None` when unknown."""
    value_try: float | None = None
    value_basis: ValueBasis = "unknown"
    weight_pct: float | None = None
    """Share of the entity's *known* lira value, as a fraction. `None` when nothing is valued."""
    source: SourceRef
    note: str | None = None
    since: str | None = None
    """Earliest daily snapshot the holder appears in for this company. With
    `at_baseline` true it means "since at least" — the real entry predates
    the first snapshot and is unknown."""
    at_baseline: bool = True
    previous_stake_pct: float | None = None
    """The stake on the previous snapshot, as a fraction. None with one snapshot."""
    delta_pct: float | None = None
    """`stake_pct - previous_stake_pct`, in fraction points. None when unknown."""


class StakeMove(BaseModel):
    """A holder entering, leaving or resizing, as read off two snapshots."""

    id: str
    ticker: str
    company: str
    holder: str
    entity_id: str | None = None
    kind: StakeMoveKind
    stake_before: float | None = None
    stake_after: float | None = None
    delta_pct: float | None = None
    observed_at: str
    """The snapshot day the change was first seen — not the filing date."""


class AllocationSlice(BaseModel):
    key: str
    label: str
    ticker: str | None = None
    value_try: float
    pct: float


class Move(BaseModel):
    id: str
    ticker: str
    company: str
    event: str
    event_label: str
    headline: str
    published_at: str | None = None
    url: str
    score: int | None = None
    band: str


class SourceHealth(BaseModel):
    kind: SourceKind
    ok: bool
    entities_covered: int = 0
    tickers_covered: int = 0
    as_of: str | None = None
    message: str | None = None


class EntitySummary(BaseModel):
    id: str
    name: str
    subtitle: str | None = None
    category: HolderCategory
    total_value_try: float | None = None
    positions_count: int = 0
    allocation: list[AllocationSlice] = Field(default_factory=list)
    top_positions: list[Position] = Field(default_factory=list)
    last_move: Move | None = None
    as_of: str | None = None
    stale: bool = False
    issues: list[str] = Field(default_factory=list)
    has_data: bool = False
    coverage_note: str | None = None


class EntityDetail(BaseModel):
    entity: EntitySummary
    positions: list[Position]
    moves: list[Move]
    stake_moves: list[StakeMove] = Field(default_factory=list)
    sources: list[SourceRef]
    tracking_since: str | None = None
    """The oldest snapshot day. Nothing before it is known."""


class Holder(BaseModel):
    """One row of a company's shareholder table, as İş Yatırım prints it."""

    label: str
    stake_pct: float
    value_try: float | None = None
    entity_id: str | None = None
    tracked: bool = False
    """Whether the row matched a registry entity. Untracked rows are still
    listed — the reader is owed every ≥5% holder, not only the ones we name."""
    since: str | None = None
    at_baseline: bool = True
    previous_stake_pct: float | None = None
    delta_pct: float | None = None


class FundHolder(BaseModel):
    entity_id: str
    name: str
    code: str
    weight_in_fund_pct: float | None = None
    """Share of the fund's equity book, as a fraction."""
    value_try: float | None = None
    stake_pct: float | None = None
    as_of: str | None = None
    url: str | None = None


class AssetOwners(BaseModel):
    ticker: str
    name: str
    market_cap: float | None = None
    free_float_pct: float | None = None
    foreign_ratio_pct: float | None = None
    """Foreign investors' share of the free float, as a fraction, via İş Yatırım."""
    holders: list[Holder] = Field(default_factory=list)
    funds: list[FundHolder] = Field(default_factory=list)
    moves: list[Move] = Field(default_factory=list)
    stake_moves: list[StakeMove] = Field(default_factory=list)
    tracking_since: str | None = None
    as_of: str | None = None
    stale: bool = False
    source_url: str | None = None


class OwnershipBoard(BaseModel):
    entities: list[EntitySummary]
    latest_moves: list[Move] = Field(default_factory=list)
    latest_stake_moves: list[StakeMove] = Field(default_factory=list)
    tracking_since: str | None = None
    category_counts: dict[str, int] = Field(default_factory=dict)
    sources: list[SourceHealth] = Field(default_factory=list)
    universe: str = "XU100"
    tickers_covered: int = 0
    tickers_total: int = 0
    as_of: str | None = None
    last_refresh_at: str | None = None
    stale: bool = False

File: frontend/lib/bist-ownership.ts
/**
 * Pure derivations for the BIST ownership board and the company-page panel.
 *
 * Kept out of the components for the reason the rest of `lib/` is: what is
 * pinned here is the handful of places where "no answer" and "zero" are
 * different things — a holder with no market cap has a stake and no value, a
 * company with no named holder is not a company nobody owns — and a
 * component that computed them inline would be tested by nobody.
 */

import { formatPercent } from '@/lib/bist-format';
import { formatDate } from '@/lib/bist-format';
import type {
  BistAssetOwners,
  BistHolderCategory,
  BistOwnershipEntity,
  BistOwnershipFacts,
  BistOwnershipSlice,
  BistOwnershipStance,
  BistStakeMoveKind,
} from '@/lib/bist-api';

export const CATEGORY_LABEL: Record<BistHolderCategory, string> = {
  state: 'Kamu',
  holding: 'Holding',
  foreign: 'Yabancı ortak',
  fund: 'Fon',
  other: 'Diğer',
};

/** Display order of the category filter. The state first: it is the largest holder on the index. */
export const CATEGORY_ORDER: BistHolderCategory[] = [
  'state',
  'holding',
  'foreign',
  'fund',
  'other',
];

export function isHolderCategory(value: string | null | undefined): value is BistHolderCategory {
  return value !== null && value !== undefined && value in CATEGORY_LABEL;
}

export function filterByCategory(
  entities: BistOwnershipEntity[],
  category: BistHolderCategory | null
): BistOwnershipEntity[] {
  if (category === null) return entities;
  return entities.filter((entity) => entity.category === category);
}

/** The pooled tail slice the server appends past the named holdings. */
export const POOLED_SLICE_KEY = '__other__';

export interface OwnerSegment {
  key: string;
  label: string;
  ticker: string | null;
  pct: number;
  /** A CSS colour expression. The pooled tail is always the muted one. */
  color: string;
  pooled: boolean;
}

/**
 * One colour per ticker, the same on every card.
 *
 * Derived from the ticker rather than from the slice's position, so THYAO is
 * the same hue on the Varlık Fonu card, the QNB fund card and the rail — a
 * reader scanning the grid can spot a company by colour before reading it.
 * A position-based palette made every first slice purple, which told the
 * reader nothing except which holding was largest, and that is already the
 * first row of the list beneath.
 *
 * The hash is FNV-1a over the code, spread across the hue wheel with a fixed
 * saturation and lightness that clear the dark surface and stay legible on
 * the light theme. Collisions between two tickers are possible and tolerable:
 * the legend names them, and two hues 20° apart on one bar are still two.
 */
export function tickerColor(ticker: string): string {
  let hash = 0x811c9dc5;
  for (let i = 0; i < ticker.length; i++) {
    hash ^= ticker.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  // Golden-angle spreading keeps neighbouring hashes far apart on the wheel,
  // and a second byte of the hash picks one of three lightness steps, so two
  // tickers that land on the same hue still differ visibly side by side.
  const hue = Math.round(((hash % 360) * 137.508) % 360);
  const lightness = [54, 62, 70][(hash >>> 9) % 3];
  return `hsl(${hue} 68% ${lightness}%)`;
}

export function allocationSegments(slices: BistOwnershipSlice[]): OwnerSegment[] {
  return slices
    .filter((slice) => slice.pct > 0)
    .map((slice) => {
      const pooled = slice.key === POOLED_SLICE_KEY;
      const color = pooled ? 'var(--fg-subtle)' : tickerColor(slice.ticker ?? slice.key);
      return {
        key: slice.key,
        label: slice.label,
        ticker: slice.ticker,
        pct: slice.pct,
        color,
        pooled,
      };
    });
}

/** One sentence for a screen reader, in place of a title per segment. */
export function allocationSummary(segments: OwnerSegment[]): string {
  if (segments.length === 0) return 'Değerlenebilen pozisyon yok';
  return segments
    .map((segment) => `${segment.label} ${formatPercent(segment.pct, segment.pct < 0.1 ? 1 : 0)}`)
    .join(', ');
}

export interface HolderCoverage {
  /** Sum of every named ≥5% stake, as a fraction of capital. */
  namedPct: number;
  /** What is left — the free float and every holder under the threshold. */
  otherPct: number;
  /** How many of the named holders the board tracks. */
  tracked: number;
  untracked: number;
}

/**
 * What the shareholder table accounts for.
 *
 * `otherPct` is derived rather than read from the card so it always agrees
 * with the rows beside it; a table listing 49% and a bucket saying 60% would
 * be the kind of arithmetic a reader checks and then stops trusting.
 */
export function holderCoverage(owners: Pick<BistAssetOwners, 'holders'>): HolderCoverage {
  const namedPct = owners.holders.reduce((sum, holder) => sum + holder.stake_pct, 0);
  const tracked = owners.holders.filter((holder) => holder.tracked).length;
  return {
    namedPct,
    otherPct: Math.max(0, 1 - namedPct),
    tracked,
    untracked: owners.holders.length - tracked,
  };
}

/**
 * What the company's page says about its ownership in one line.
 *
 * Three sentences, because they are three facts: nobody crosses the
 * disclosure threshold, one holder controls the company outright, or the
 * capital is split between named holders.
 */
export function holderHeadline(owners: Pick<BistAssetOwners, 'holders'>): string {
  if (owners.holders.length === 0) {
    return '%5 eşiğini geçen ortak yok; sermayenin tamamı halka açık ve küçük paylardan oluşuyor.';
  }
  const largest = owners.holders[0];
  if (largest.stake_pct > 0.5) {
    return `${largest.label} ${formatPercent(largest.stake_pct)} ile çoğunluk ortağı.`;
  }
  const { namedPct } = holderCoverage(owners);
  return `${owners.holders.length} ortak sermayenin ${formatPercent(namedPct)}'ini tutuyor; en büyüğü ${largest.label} (${formatPercent(largest.stake_pct)}).`;
}

export interface BoardSummary {
  entities: number;
  withData: number;
  /** Sum of every entity's known lira value. `null` when nothing was valued. */
  totalValued: number | null;
  /** How many cards came back, over how many were asked for. */
  coverage: string;
}

export function boardSummary(board: {
  entities: BistOwnershipEntity[];
  tickers_covered: number;
  tickers_total: number;
  universe: string;
}): BoardSummary {
  const withData = board.entities.filter((entity) => entity.has_data);
  const valued = withData
    .map((entity) => entity.total_value_try)
    .filter((v): v is number => v !== null);
  return {
    entities: board.entities.length,
    withData: withData.length,
    totalValued: valued.length ? valued.reduce((sum, v) => sum + v, 0) : null,
    coverage: `${board.tickers_covered}/${board.tickers_total} ${board.universe}`,
  };
}

/** The Turkish word for how a lira value was arrived at. */
export const VALUE_BASIS_LABEL = {
  marked: 'piyasa değeriyle',
  reported: 'raporlanan',
  unknown: 'değerlenemedi',
} as const;

export const STAKE_MOVE_LABEL: Record<BistStakeMoveKind, string> = {
  new: 'Giriş',
  exit: 'Çıkış',
  add: 'Artış',
  trim: 'Azalış',
};

/**
 * When a holder came in, as far as the snapshots can tell.
 *
 * Three sentences for three states of knowledge: no snapshot has seen the
 * holder yet; the holder was already there on the first snapshot, so the
 * entry predates what is known; or the holder appeared on a later day, which
 * is the only case the word "giriş" is used for. Printing the baseline day as
 * an entry date would make every holder on the index look like it arrived
 * the day this feature shipped.
 */
export function sinceLabel(since: string | null, atBaseline: boolean): string {
  if (since === null) return 'Kayıt yok';
  if (atBaseline) return `≤ ${formatDate(since)}`;
  return `Giriş ${formatDate(since)}`;
}

/**
 * A stake change in points of capital, signed, or what stands in for one.
 *
 * `null` is "unknown" (one snapshot, nothing to compare) and is printed as a
 * dash; `0` is "unchanged" and is printed as such. The two must not share a
 * glyph: on the day after the first snapshot every row is genuinely
 * unchanged, and a dash there would read as the feature not working.
 */
export function formatStakeDelta(delta: number | null): string {
  if (delta === null) return '—';
  if (Math.abs(delta) < 0.0001) return '0';
  const points = delta * 100;
  const sign = points > 0 ? '+' : '';
  return `${sign}${points.toFixed(Math.abs(points) < 1 ? 2 : 1)} puan`;
}

export const OWNERSHIP_STANCE_LABEL: Record<BistOwnershipStance, string> = {
  state_anchored: 'Endeksin çapası kamu',
  family_holdings: 'Endeksin çapası holdingler',
  foreign_strategic: 'Endeksin çapası yabancı stratejik ortaklar',
  dispersed: 'Sahiplik dağınık',
};

export const OWNERSHIP_STANCE_TONE: Record<BistOwnershipStance, string> = {
  state_anchored: 'text-fg',
  family_holdings: 'text-fg',
  foreign_strategic: 'text-fg',
  dispersed: 'text-fg-muted',
};

export interface OwnershipChip {
  text: string;
  title: string;
}

/**
 * The three or four readings that decided the stance, as compact chips beside
 * it. Rendered from the facts alone, like the prompt, so the chips and the
 * paragraph can never disagree about a figure.
 */
export function ownershipChips(facts: BistOwnershipFacts): OwnershipChip[] {
  const chips: OwnershipChip[] = [];
  const [first, second] = facts.total.categories;
  if (first && first.share_pct !== null) {
    chips.push({
      title: 'Değerlenen toplamın en büyük iki ortak türüne düşen payı',
      text:
        `${CATEGORY_LABEL[first.category].toLocaleLowerCase('tr')} %${first.share_pct}` +
        (second && second.share_pct !== null
          ? ` · ${CATEGORY_LABEL[second.category].toLocaleLowerCase('tr')} %${second.share_pct}`
          : ''),
    });
  }
  if (facts.holders.top3_share_pct !== null) {
    chips.push({
      title: 'En büyük üç ortağın değerlenen toplamdaki payı',
      text: `ilk 3 ortak %${facts.holders.top3_share_pct}`,
    });
  }
  if (facts.companies.median_named_stake_pct !== null) {
    chips.push({
      title: 'Medyan şirkette %5 üzeri ortakların tuttuğu toplam pay',
      text: `medyan sahiplik %${facts.companies.median_named_stake_pct}`,
    });
  }
  if (facts.companies.median_foreign_ratio_pct !== null) {
    chips.push({
      title: 'Halka açık kısımdaki yabancı payı, XU100 medyanı',
      text: `yabancı medyan %${facts.companies.median_foreign_ratio_pct}`,
    });
  }
  chips.push({
    title: 'Şirketlerden kaçında bir ortak sermayenin yarısından fazlasını tutuyor',
    text: `${facts.companies.majority_held} şirkette çoğunluk ortağı`,
  });
  return chips;
}
